File: implementations/cpp/wavelet_trees.cpp
/**
 * @file wavelet_trees.cpp
 * @brief High-performance C++17 implementation of Wavelet Trees.
 *
 * Implements Grossi, Gupta, and Vitter's succinct sequence decomposition (SODA 2003):
 * - BitVector with packed 64-bit words and prefix popcount directory for O(1) rank queries.
 * - Wavelet Tree over arbitrary integer alphabet Sigma:
 *   1. Access(i): Retrieve A[i] in O(log |Sigma|).
 *   2. Rank(i, c): Count occurrences of c in A[0 ... i] in O(log |Sigma|).
 *   3. RangeCount(L, R, c): Count occurrences of c in A[L ... R] in O(log |Sigma|).
 *   4. Quantile(L, R, k): Find k-th smallest element in A[L ... R] in O(log |Sigma|).
 *   5. RangeFrequency(L, R, low, high): Count elements in A[L ... R] with values in [low, high] in O(log |Sigma|).
 * - Total Space: O(N log |Sigma|) bits = O(N) words.
 *
 * Fully compliant with -std=c++17 -O3 -Wall -Wextra -Werror.
 */

#include <iostream>
#include <vector>
#include <cstdint>
#include <cassert>
#include <algorithm>
#include <random>

namespace dsa {

/**
 * @brief BitVector supporting O(1) rank queries using 64-bit blocks and prefix directory.
 */
class BitVector {
private:
    std::vector<uint64_t> words_;
    std::vector<uint32_t> prefix_popcounts_;
    size_t n_{0};

public:
    BitVector() = default;

    explicit BitVector(const std::vector<bool>& bits) : n_(bits.size()) {
        size_t num_words = (n_ + 63) / 64;
        words_.assign(num_words, 0);
        prefix_popcounts_.assign(num_words + 1, 0);

        for (size_t i = 0; i < n_; ++i) {
            if (bits[i]) {
                words_[i / 64] |= (1ULL << (i % 64));
            }
        }

        uint32_t running = 0;
        for (size_t i = 0; i < num_words; ++i) {
            prefix_popcounts_[i] = running;
            running += __builtin_popcountll(words_[i]);
        }
        prefix_popcounts_[num_words] = running;
    }

    [[nodiscard]] size_t size() const noexcept { return n_; }

    [[nodiscard]] bool get(size_t i) const noexcept {
        return (words_[i / 64] >> (i % 64)) & 1ULL;
    }

    /**
     * @brief Number of 1-bits in prefix [0, i].
     */
    [[nodiscard]] uint32_t rank1(int64_t i) const noexcept {
        if (i < 0) return 0;
        if (static_cast<size_t>(i) >= n_) i = n_ - 1;

        size_t word_idx = static_cast<size_t>(i) / 64;
        size_t bit_idx = static_cast<size_t>(i) % 64;

        uint32_t count = prefix_popcounts_[word_idx];
        uint64_t mask = (bit_idx == 63) ? ~0ULL : ((1ULL << (bit_idx + 1)) - 1ULL);
        count += __builtin_popcountll(words_[word_idx] & mask);
        return count;
    }

    /**
     * @brief Number of 0-bits in prefix [0, i].
     */
    [[nodiscard]] uint32_t rank0(int64_t i) const noexcept {
        if (i < 0) return 0;
        if (static_cast<size_t>(i) >= n_) i = n_ - 1;
        return static_cast<uint32_t>(i + 1) - rank1(i);
    }
};

/**
 * @brief Wavelet Tree over an arbitrary sequence of integers in [min_val, max_val].
 */
class WaveletTree {
private:
    int64_t low_{0};
    int64_t high_{0};
    BitVector bv_;
    WaveletTree* left_{nullptr};
    WaveletTree* right_{nullptr};

public:
    WaveletTree() = default;

    template <typename It>
    WaveletTree(It from, It to, int64_t low, int64_t high) : low_(low), high_(high) {
        if (from >= to || low_ >= high_) return;

        int64_t mid = low_ + (high_ - low_) / 2;
        std::vector<bool> bits;
        bits.reserve(to - from);

        for (auto it = from; it != to; ++it) {
            bits.push_back(*it > mid);
        }
        bv_ = BitVector(bits);

        // Partition elements stably into left (<= mid) and right (> mid)
        auto pivot = std::stable_partition(from, to, [mid](int64_t x) {
            return x <= mid;
        });

        left_ = new WaveletTree(from, pivot, low_, mid);
        right_ = new WaveletTree(pivot, to, mid + 1, high_);
    }

    ~WaveletTree() {
        delete left_;
        delete right_;
    }

    WaveletTree(const WaveletTree&) = delete;
    WaveletTree& operator=(const WaveletTree&) = delete;

    WaveletTree(WaveletTree&& other) noexcept
        : low_(other.low_), high_(other.high_), bv_(std::move(other.bv_)),
          left_(other.left_), right_(other.right_) {
        other.left_ = other.right_ = nullptr;
    }

    WaveletTree& operator=(WaveletTree&& other) noexcept {
        if (this != &other) {
            delete left_;
            delete right_;
            low_ = other.low_;
            high_ = other.high_;
            bv_ = std::move(other.bv_);
            left_ = other.left_;
            right_ = other.right_;
            other.left_ = other.right_ = nullptr;
        }
        return *this;
    }

    /**
     * @brief Access element at index i (0-indexed).
     */
    [[nodiscard]] int64_t access(size_t i) const {
        if (low_ == high_) return low_;
        bool bit = bv_.get(i);
        if (!bit) {
            return left_->access(bv_.rank0(static_cast<int64_t>(i)) - 1);
        } else {
            return right_->access(bv_.rank1(static_cast<int64_t>(i)) - 1);
        }
    }

    /**
     * @brief Count occurrences of symbol c in prefix [0, i].
     */
    [[nodiscard]] uint32_t rank(int64_t i, int64_t c) const {
        if (i < 0 || c < low_ || c > high_) return 0;
        if (low_ == high_) return static_cast<uint32_t>(i + 1);

        int64_t mid = low_ + (high_ - low_) / 2;
        if (c <= mid) {
            return left_->rank(static_cast<int64_t>(bv_.rank0(i)) - 1, c);
        } else {
            return right_->rank(static_cast<int64_t>(bv_.rank1(i)) - 1, c);
        }
    }

    /**
     * @brief Count occurrences of symbol c in range [L, R].
     */
    [[nodiscard]] uint32_t range_count(size_t L, size_t R, int64_t c) const {
        if (L > R) return 0;
        return rank(static_cast<int64_t>(R), c) - rank(static_cast<int64_t>(L) - 1, c);
    }

    /**
     * @brief Find k-th smallest element (1-indexed) in range [L, R].
     */
    [[nodiscard]] int64_t quantile(size_t L, size_t R, uint32_t k) const {
        if (low_ == high_) return low_;

        int64_t left_L = static_cast<int64_t>(bv_.rank0(static_cast<int64_t>(L) - 1));
        int64_t left_R = static_cast<int64_t>(bv_.rank0(static_cast<int64_t>(R))) - 1;
        uint32_t zeros_in_range = (left_R >= left_L) ? static_cast<uint32_t>(left_R - left_L + 1) : 0;

        if (k <= zeros_in_range) {
            return left_->quantile(static_cast<size_t>(left_L), static_cast<size_t>(left_R), k);
        } else {
            int64_t right_L = static_cast<int64_t>(bv_.rank1(static_cast<int64_t>(L) - 1));
            int64_t right_R = static_cast<int64_t>(bv_.rank1(static_cast<int64_t>(R))) - 1;
            return right_->quantile(static_cast<size_t>(right_L), static_cast<size_t>(right_R), k - zeros_in_range);
        }
    }

    /**
     * @brief Count number of elements in range [L, R] whose values lie in [val_low, val_high].
     */
    [[nodiscard]] uint32_t range_frequency(size_t L, size_t R, int64_t val_low, int64_t val_high) const {
        if (L > R || val_low > high_ || val_high < low_) return 0;
        if (val_low <= low_ && high_ <= val_high) {
            return static_cast<uint32_t>(R - L + 1);
        }

        int64_t left_L = static_cast<int64_t>(bv_.rank0(static_cast<int64_t>(L) - 1));
        int64_t left_R = static_cast<int64_t>(bv_.rank0(static_cast<int64_t>(R))) - 1;

        int64_t right_L = static_cast<int64_t>(bv_.rank1(static_cast<int64_t>(L) - 1));
        int64_t right_R = static_cast<int64_t>(bv_.rank1(static_cast<int64_t>(R))) - 1;

        uint32_t res = 0;
        if (left_R >= left_L && left_) {
            res += left_->range_frequency(static_cast<size_t>(left_L), static_cast<size_t>(left_R), val_low, val_high);
        }
        if (right_R >= right_L && right_) {
            res += right_->range_frequency(static_cast<size_t>(right_L), static_cast<size_t>(right_R), val_low, val_high);
        }
        return res;
    }
};

} // namespace dsa

int main() {
    using namespace dsa;
    std::cout << "Testing Wavelet Tree against brute-force oracles...\n";

    // 1. Basic operations verification
    {
        std::vector<int64_t> arr = {5, 2, 8, 3, 9, 2, 7, 1, 6, 4, 3, 8};
        std::vector<int64_t> copy_arr = arr;
        WaveletTree wt(copy_arr.begin(), copy_arr.end(), 1, 9);

        for (size_t i = 0; i < arr.size(); ++i) {
            assert(wt.access(i) == arr[i]);
        }

        assert(wt.range_count(0, 11, 2) == 2);
        assert(wt.range_count(0, 11, 8) == 2);
        assert(wt.range_count(0, 11, 10) == 0);

        assert(wt.quantile(0, 4, 1) == 2); // {5, 2, 8, 3, 9} -> sorted: {2, 3, 5, 8, 9}
        assert(wt.quantile(0, 4, 2) == 3);
        assert(wt.quantile(0, 4, 3) == 5);
        assert(wt.quantile(0, 4, 5) == 9);

        assert(wt.range_frequency(0, 4, 2, 5) == 3); // 2, 3, 5
    }

    // 2. Large Scale Randomized Differential Verification
    {
        std::mt19937 rng(1337);
        size_t N = 1000;
        std::vector<int64_t> big_arr(N);
        std::uniform_int_distribution<int64_t> val_dist(0, 200);
        for (size_t i = 0; i < N; ++i) {
            big_arr[i] = val_dist(rng);
        }
        std::vector<int64_t> big_copy = big_arr;
        WaveletTree big_wt(big_copy.begin(), big_copy.end(), 0, 200);

        for (int it = 0; it < 5000; ++it) {
            size_t l = rng() % N;
            size_t r = rng() % N;
            if (l > r) std::swap(l, r);

            // Quantile check
            uint32_t k = (rng() % (r - l + 1)) + 1;
            std::vector<int64_t> sub(big_arr.begin() + l, big_arr.begin() + r + 1);
            std::sort(sub.begin(), sub.end());
            int64_t expected_kth = sub[k - 1];
            int64_t actual_kth = big_wt.quantile(l, r, k);
            assert(actual_kth == expected_kth);

            // Range frequency check
            int64_t v1 = val_dist(rng);
            int64_t v2 = val_dist(rng);
            if (v1 > v2) std::swap(v1, v2);

            uint32_t expected_freq = 0;
            for (size_t i = l; i <= r; ++i) {
                if (big_arr[i] >= v1 && big_arr[i] <= v2) expected_freq++;
            }
            uint32_t actual_freq = big_wt.range_frequency(l, r, v1, v2);
            assert(actual_freq == expected_freq);

            // Range count check
            int64_t c = val_dist(rng);
            uint32_t expected_c = 0;
            for (size_t i = l; i <= r; ++i) {
                if (big_arr[i] == c) expected_c++;
            }
            assert(big_wt.range_count(l, r, c) == expected_c);
        }
    }

    std::cout << "Wavelet Tree all tests passed with 100% precision!\n";
    return 0;
}

File: implementations/python/wavelet_trees.py
"""
Wavelet Tree Implementation in Python 3.

Provides:
- BitVector: Succinct bitvector supporting O(1) rank queries using 64-bit blocks.
- WaveletTree: Succinct sequence representation over an alphabet Sigma, supporting:
  - access(i): O(log |Sigma|)
  - rank(i, c): O(log |Sigma|)
  - range_count(L, R, c): O(log |Sigma|)
  - quantile(L, R, k): O(log |Sigma|)
  - range_frequency(L, R, low, high): O(log |Sigma|)
"""

import random
import unittest
from typing import List, Sequence, Optional


class BitVector:
    """
    BitVector supporting O(1) rank queries via 64-bit packed words and prefix directory.
    """
    def __init__(self, bits: Sequence[bool]):
        self.n = len(bits)
        num_words = (self.n + 63) // 64
        self.words = [0] * num_words
        self.prefix_popcounts = [0] * (num_words + 1)

        for i, b in enumerate(bits):
            if b:
                self.words[i // 64] |= (1 << (i % 64))

        running = 0
        for i in range(num_words):
            self.prefix_popcounts[i] = running
            running += bin(self.words[i]).count('1')
        self.prefix_popcounts[num_words] = running

    def __len__(self) -> int:
        return self.n

    def get(self, i: int) -> bool:
        return bool((self.words[i // 64] >> (i % 64)) & 1)

    def rank1(self, i: int) -> int:
        """Count 1-bits in prefix [0, i]."""
        if i < 0:
            return 0
        if i >= self.n:
            i = self.n - 1
        word_idx = i // 64
        bit_idx = i % 64
        count = self.prefix_popcounts[word_idx]
        mask = (1 << (bit_idx + 1)) - 1
        count += bin(self.words[word_idx] & mask).count('1')
        return count

    def rank0(self, i: int) -> int:
        """Count 0-bits in prefix [0, i]."""
        if i < 0:
            return 0
        if i >= self.n:
            i = self.n - 1
        return (i + 1) - self.rank1(i)


class WaveletTree:
    """
    Wavelet Tree over integer sequence with alphabet range [low, high].
    """
    def __init__(self, arr: Sequence[int], low: int, high: int):
        self.low = low
        self.high = high
        self.left: Optional['WaveletTree'] = None
        self.right: Optional['WaveletTree'] = None

        if not arr or low >= high:
            self.bv = BitVector([])
            return

        mid = low + (high - low) // 2
        bits = [x > mid for x in arr]
        self.bv = BitVector(bits)

        left_arr = [x for x in arr if x <= mid]
        right_arr = [x for x in arr if x > mid]

        self.left = WaveletTree(left_arr, low, mid)
        self.right = WaveletTree(right_arr, mid + 1, high)

    def access(self, i: int) -> int:
        """Retrieve element at index i (0-indexed)."""
        if self.low == self.high:
            return self.low
        bit = self.bv.get(i)
        if not bit:
            assert self.left is not None
            return self.left.access(self.bv.rank0(i) - 1)
        else:
            assert self.right is not None
            return self.right.access(self.bv.rank1(i) - 1)

    def rank(self, i: int, c: int) -> int:
        """Count occurrences of symbol c in prefix [0, i]."""
        if i < 0 or c < self.low or c > self.high:
            return 0
        if self.low == self.high:
            return i + 1

        mid = self.low + (self.high - self.low) // 2
        if c <= mid:
            assert self.left is not None
            return self.left.rank(self.bv.rank0(i) - 1, c)
        else:
            assert self.right is not None
            return self.right.rank(self.bv.rank1(i) - 1, c)

    def range_count(self, L: int, R: int, c: int) -> int:
        """Count occurrences of symbol c in range [L, R]."""
        if L > R:
            return 0
        return self.rank(R, c) - self.rank(L - 1, c)

    def quantile(self, L: int, R: int, k: int) -> int:
        """Find k-th smallest element in range [L, R] (1-indexed)."""
        if self.low == self.high:
            return self.low

        left_L = self.bv.rank0(L - 1)
        left_R = self.bv.rank0(R) - 1
        zeros_in_range = max(0, left_R - left_L + 1)

        if k <= zeros_in_range:
            assert self.left is not None
            return self.left.quantile(left_L, left_R, k)
        else:
            assert self.right is not None
            right_L = self.bv.rank1(L - 1)
            right_R = self.bv.rank1(R) - 1
            return self.right.quantile(right_L, right_R, k - zeros_in_range)

    def range_frequency(self, L: int, R: int, val_low: int, val_high: int) -> int:
        """Count number of elements in range [L, R] whose values lie in [val_low, val_high]."""
        if L > R or val_low > self.high or val_high < self.low:
            return 0
        if val_low <= self.low and self.high <= val_high:
            return R - L + 1

        left_L = self.bv.rank0(L - 1)
        left_R = self.bv.rank0(R) - 1

        right_L = self.bv.rank1(L - 1)
        right_R = self.bv.rank1(R) - 1

        res = 0
        if left_R >= left_L and self.left:
            res += self.left.range_frequency(left_L, left_R, val_low, val_high)
        if right_R >= right_L and self.right:
            res += self.right.range_frequency(right_L, right_R, val_low, val_high)
        return res


class TestWaveletTree(unittest.TestCase):
    def test_basic_queries(self):
        arr = [5, 2, 8, 3, 9, 2, 7, 1, 6, 4, 3, 8]
        wt = WaveletTree(arr, 1, 9)

        for i, val in enumerate(arr):
            self.assertEqual(wt.access(i), val)

        self.assertEqual(wt.range_count(0, 11, 2), 2)
        self.assertEqual(wt.range_count(0, 11, 8), 2)
        self.assertEqual(wt.range_count(0, 11, 10), 0)

        # {5, 2, 8, 3, 9} -> sorted: {2, 3, 5, 8, 9}
        self.assertEqual(wt.quantile(0, 4, 1), 2)
        self.assertEqual(wt.quantile(0, 4, 2), 3)
        self.assertEqual(wt.quantile(0, 4, 3), 5)
        self.assertEqual(wt.quantile(0, 4, 5), 9)

        self.assertEqual(wt.range_frequency(0, 4, 2, 5), 3)

    def test_random_differential(self):
        rng = random.Random(42)
        N = 300
        arr = [rng.randint(0, 50) for _ in range(N)]
        wt = WaveletTree(arr, 0, 50)

        for _ in range(1500):
            l = rng.randint(0, N - 1)
            r = rng.randint(0, N - 1)
            if l > r:
                l, r = r, l

            sub = arr[l:r + 1]

            # Quantile
            k = rng.randint(1, len(sub))
            sorted_sub = sorted(sub)
            self.assertEqual(wt.quantile(l, r, k), sorted_sub[k - 1])

            # Range count
            c = rng.randint(0, 50)
            self.assertEqual(wt.range_count(l, r, c), sub.count(c))

            # Range frequency
            v1 = rng.randint(0, 50)
            v2 = rng.randint(0, 50)
            if v1 > v2:
                v1, v2 = v2, v1
            expected_freq = sum(1 for x in sub if v1 <= x <= v2)
            self.assertEqual(wt.range_frequency(l, r, v1, v2), expected_freq)


if __name__ == '__main__':
    unittest.main()
